### PR TITLE
[SYCL] Change check_device_code CUDA tests to use SYCL_EXTERNAL

### PR DESCRIPTION
**Pull request by ianayl**
Changed CUDA sycl/test/check_device_code lit test cases to use SYCL_EXTERNAL functions instead of submitting kernels to the queue everytime.

### CODE REVIEW

File: sycl/test/check_device_code/cuda/ldg.cpp [hunk 1]
```diff
@@ -9,279 +10,509 @@ using namespace sycl;
 using namespace sycl::ext::oneapi::experimental::cuda;
 using namespace sycl::ext::oneapi::experimental;
 
-int main() {
+// CHECK-OPAQUE-LABEL: define weak_odr dso_local %"class.sycl::_V1::detail::half_impl::half" @_ZN4sycl3_V13ext6oneapi12experimental4cuda3ldgINS0_6detail9half_impl4halfEEENSt9enable_ifIXsr4sycl6detail12is_containedIT_NS6_5boost4mp117mp_listIJNS0_3vecIcLi2EEENSE_IcLi3EEENSE_IcLi4EEENSE_IaLi2EEENSE_IaLi3EEENSE_IaLi4EEENSE_IsLi2EEENSE_IsLi3EEENSE_IsLi4EEENSE_IiLi2EEENSE_IiLi3EEENSE_IiLi4EEENSE_IlLi2EEENSE_IlLi3EEENSE_IlLi4EEENSE_IxLi2EEENSE_IxLi3EEENSE_IxLi4EEENSE_IhLi2EEENSE_IhLi3EEENSE_IhLi4EEENSE_ItLi2EEENSE_ItLi3EEENSE_ItLi4EEENSE_IjLi2EEENSE_IjLi3EEENSE_IjLi4EEENSE_ImLi2EEENSE_ImLi3EEENSE_ImLi4EEENSE_IyLi2EEENSE_IyLi3EEENSE_IyLi4EEENSE_IS8_Li2EEENSE_IS8_Li3EEENSE_IS8_Li4EEENSE_IfLi2EEENSE_IfLi3EEENSE_IfLi4EEENSE_IdLi2EEENSE_IdLi3EEENSE_IdLi4EEEfdS8_NS2_8bfloat16EcasilxhtjmySt4byteEEEEE5valueESA_E4typeEPKSA_(
```
Comment (JackAKirk) [inline sycl/test/check_device_code/cuda/ldg.cpp:13]: I think that all this checking is probably overkill, the purpose of the test is really to only check these parts:
```
tail call half @llvm.nvvm.ldg.global.f.f16.p0
```

The other stuff can (and sometimes does) change, but is orthogonal to the purposes of these tests.

Could you update it to be more consistent with how the checks were previously, which were more focused on checking the correct llvm builtins are being generated?

**Comment by ianayl** [comment]
> Please revert the permission changes in buildbot/configure.py.
> 
> @JackAKirk - Would you mind having a look at this one?

My bad, that was accidental! I've fixed it now.

File: sycl/test/check_device_code/cuda/ldg.cpp (hunk 1 above)
Comment (ianayl) [reply sycl/test/check_device_code/cuda/ldg.cpp]: That makes sense, I've gone ahead and removed all the other tests.

File: sycl/test/check_device_code/cuda/ldg.cpp [hunk 2]
```diff
@@ -9,279 +10,146 @@ using namespace sycl;
...
+template SYCL_EXTERNAL sycl::vec<short, 4> sycl::ext::oneapi::experimental::cuda::ldg(const sycl::vec<short, 4> *);
+// CHECK-OPAQUE: tail call <4 x i32> @llvm.nvvm.ldg.global.i.v4i32.p0(ptr %{{.*}}, i32 16)
+template SYCL_EXTERNAL sycl::vec<int, 4> sycl::ext::oneapi::experimental::cuda::ldg(const sycl::vec<int, 4> *);
+
+// CHECK-OPAQUE: tail call <4 x i8> @llvm.nvvm.ldg.global.i.v4i8.p0(ptr %{{.*}}, i32 4)
+template SYCL_EXTERNAL sycl::vec<unsigned char, 4> sycl::ext::oneapi::experimental::cuda::ldg(const sycl::vec<unsigned char, 4> *);
+// CHECK-OPAQUE: tail call <4 x i16> @llvm.nvvm.ldg.global.i.v4i16.p0(ptr %{{.*}}, i32 8)
+template SYCL_EXTERNAL sycl::vec<unsigned short, 4> sycl::ext::oneapi::experimental::cuda::ldg(const sycl::vec<unsigned short, 4> *);
+// CHECK-OPAQUE: tail call <4 x i32> @llvm.nvvm.ldg.global.i.v4i32.p0(ptr %{{.*}}, i32 16)
+template SYCL_EXTERNAL sycl::vec<unsigned int, 4> sycl::ext::oneapi::experimental::cuda::ldg(const sycl::vec<unsigned int, 4> *);
```
Comment (JackAKirk) [inline sycl/test/check_device_code/cuda/ldg.cpp:155]: missing eof here

File: sycl/test/check_device_code/cuda/ldg.cpp (hunk 1 above)
Comment (JackAKirk) [reply sycl/test/check_device_code/cuda/ldg.cpp]: Thanks, it looks good. Most of the files are missing EOF. If you can fix that then we can run the testing on this.

File: sycl/test/check_device_code/cuda/ldg.cpp (hunk 2 above)
Comment (JackAKirk) [reply sycl/test/check_device_code/cuda/ldg.cpp]: Yeah just a new line.

Comment (ianayl) [reply sycl/test/check_device_code/cuda/ldg.cpp]: I see! I did just run the the formatter and it changed as such. Is there anything else I should change?

Thanks for reviewing by the way!

Comment (JackAKirk) [reply sycl/test/check_device_code/cuda/ldg.cpp]: No problem. Yeah so long as the tests pass, this is fine now.

**Comment by ianayl** [comment]
@steffenlarsen Please excuse the ping, is there anything else I need to do for this PR to merge?

Thanks!

File: sycl/test/check_device_code/cuda/matrix/matrix-nvptx-half-float-test.cpp [hunk 3]
```diff
@@ -9,215 +9,230 @@ using namespace sycl::ext::oneapi::experimental::matrix;
...
-          // CHECK-OPAQUE: tail call void @llvm.nvvm.wmma.m8n32k16.store.d.col.stride.f32.p1(ptr addrspace(1) %{{.*}}, float {{.*}}, float {{.*}}, float {{.*}}, float {{.*}}, float {{.*}}, float {{.*}}, float {{.*}}, float {{.*}}, i32 16)
-          joint_matrix_store(
-              sg, sub_c, accD.template get_multi_ptr<access::decorated::yes>(),
-              stride, layout::col_major);
-        });
-  });
-
-  return 0;
-};
+SYCL_EXTERNAL [[sycl::reqd_work_group_size(1, 1, 32)]] void
```
Comment (steffenlarsen) [inline sycl/test/check_device_code/cuda/matrix/matrix-nvptx-half-float-test.cpp:12]: Woaw, these tests were weird. A 3-dimensional required work-group size and a 2-dimensional `nd_item` argument? It's clear that these were generated code check only.

Not a comment on the PR, just an observation.

**Review by steffenlarsen — APPROVED** [review]
LGTM, though I would like @JackAKirk to give the final approval.

**Review by JackAKirk — APPROVED** [review]

**Pull request merged**